### PR TITLE
[Java][Client][vert.x] Support per-call authentication and JsonNullable in client request bodies, fix path parameter encoding

### LINKED ISSUE #6428: [BUG][Java][Vert.x] `JsonNullable` in request body model class generates invalid requests
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

If a request model class uses a optional nullable field it get's wrapped in a `JsonNullable<T>`. This is perfectly fine. But the vert.x client then uses an `ObjectMapper` instance which lacks the registration of the `JsonNullableModule` required for the intended behaviour of differentiation between `null` and `absent`.

##### openapi-generator version

4.x, master

##### Related issues/PRs

#5732

### LINKED ISSUE #6876: [BUG][Java][Vert.x] Path parameters are not properly url encoded
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

The generated vert.x web client code does not url encode path parameters and but they are simply put as is into the url path. So whenever the parameter contains incompatible characters the URL gets messed up.

##### openapi-generator version

latest master

##### Related issues/PRs

#5732

##### Suggest a fix

Just URL encode all path parameters before adding them to the path.

### PR DESCRIPTION
**Pull request by jekkel**
Extensions
----------
Since vertx is an asynchronous/reactive toolkit usage of mutable fields of `ApiClient` to store authentication prohibits delegation of authentication from incoming call without recreating the client
everytime (which is prohibitively expensive due to instantiation of objectmapper). This commit adds a per call authentication override which takes precedence over the mutable fields so a single client can be safely reused. To ease usage a simple builder-style auxiliary `AuthInfo` object is provided providing methods for each specified authentication.

A new configuration option for `ApiClient` has been added: `timeout`. It gets applied as timeout for the vertx http request.

Fixes #6428
-----
Request bodies have been wrongly serialized using vert.x built-in `ObjectMapper` instance instead of the embedded instance thus usage of beans with `JsonNullable` wrapped fields in request bodies led to bad requests because the required Jackson module is not registered in the vert.x mapper instance.

Fixes #6876 
-----
Path parameters have been replaced as is in the operation path template. This breaks whenever a parameter contains reserved characters. Now all path parameters are properly URL encoded prior to replacement.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@bbdouglas
@sreeshas
@jfiala
@lukoyanov
@cbornet
@jeff9finger
@karismann
@Zomzog
@lwlee2608
@bkabrda

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/Java/libraries/vertx/api.mustache [hunk 1]
```diff
@@ -14,6 +15,8 @@ public interface {{classname}} {
     {{#operation}}
     void {{operationId}}({{#allParams}}{{{dataType}}} {{paramName}}, {{/allParams}}Handler<AsyncResult<{{#returnType}}{{{returnType}}}{{/returnType}}{{^returnType}}Void{{/returnType}}>> handler);
 
+    void {{operationId}}({{#allParams}}{{{dataType}}} {{paramName}}, {{/allParams}}ApiClient.AuthInfo authInfo, Handler<AsyncResult<{{#returnType}}{{{returnType}}}{{/returnType}}{{^returnType}}Void{{/returnType}}>> handler);
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/Java/libraries/vertx/api.mustache:18]: @jekkel Looks like a breaking change to me as the method signature has changed.

Comment (jekkel) [reply modules/openapi-generator/src/main/resources/Java/libraries/vertx/api.mustache:18]: Thanks for reviewing. No, it's not a breaking change as it just adds an overload with a changed signature. The previous signature is still available.

**Comment by jekkel** [comment]
@wing328 Is there anything I can assist with so this gets merged?

**Comment by jekkel** [comment]
@bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog @lwlee2608 @bkabrda

This PR has been open for about two months and I just made a not so trivial rebase onto master (conflicts in generated files gave me a little headache). Is there anything I can do so this gets merged? Is the purpose unclear so should I explain more? Do you disagree with the changeset in any other way? Please just let me know, I can deal with that. I'd be very disappointed if this hung around for another 2 month period without any activity. :cry:

**Comment by wing328** [comment]
@jekkel sorry about that :( We'd too many PRs. I'll try to review today or later this week.

**Pull request merged**